taxjar/taxjar-api-docs
# INT-837 Automate Product Tax Category Updates

**Context**
TaxJar periodically updates product tax code descriptions and we'd like to keep the API reference up to date with what's returned from the API on a continual basis while minimizing back and forth communication.

**What Changed?**
We're adding a GitHub Actions workflow to run a daily cron job that compares the [product tax categories from the API](https://api.taxjar.com/v2/categories) with the [product tax categories listed on our API reference](https://developers.taxjar.com/api/reference/#get-list-tax-categories).

If there are any differences, a diff will be shown while the job is running:

<img width="1100" alt="Screen Shot 2020-05-11 at 4 19 49 PM" src="https://user-images.githubusercontent.com/26824724/81621457-48e2d900-93a3-11ea-9e01-4c5ddfabd34b.png">

And `source/includes/endpoints/_categories.md` will be automatically synced with the product tax categories returned from the API.

If there are any new categories (not just a description change), `source/includes/_changelog.md` will be updated to display the new product tax codes in our API reference [here](https://developers.taxjar.com/api/reference/#changelog):

<img width="835" alt="Screen Shot 2020-05-11 at 4 34 11 PM" src="https://user-images.githubusercontent.com/26824724/81622270-60bb5c80-93a5-11ea-939f-0a03e44a6dcc.png">

Finally, an automated pull request will be opened for real-life people to review and approve before merging. Here's a sample:

<img width="1028" alt="Screen Shot 2020-05-11 at 4 12 24 PM" src="https://user-images.githubusercontent.com/26824724/81621715-fb1aa080-93a3-11ea-88a0-f328cc01dae0.png">

**Performance**
This change should not affect performance for https://developers.taxjar.com. However, we should ensure that none of the files added in this PR are deployed with the site.

**Testing**
There's a few different ways to test:

- View [this test commit](https://github.com/taxjar/taxjar-api-docs/commit/9278698886edaa9ed35361716d56f7f9d4e832de) to see what changed with a local run of the script. (The diff for `_categories.md` is quite long because of minor differences currently in place.)

- Alternatively, if you'd like to see the GitHub Action in action, clone this repo into a separate, personal (optionally private) repo:
  - `git clone --single-branch --branch automate-ptc-updates`
  - Change the following in `.github/workflows/sync_categories.yml`:
```yml
on:
  schedule:
    - cron: '0 23 * * *' # 11 PM UTC, 3 or 4 PM Pacific (depending on daylight savings)
```

to `'*/5 * * * *'` to run the workflow every 5 mins and wait.

Or, change to the following and push to master to trigger the job:

```yml
on:
  push:
    branches:
      - master
```
  - You may also test different scenarios, such as:
    - Faking new categories (see [test commit](https://github.com/taxjar/taxjar-api-docs/commit/9278698886edaa9ed35361716d56f7f9d4e832de))
    - Remove a category from `apiCats` and then ensure the job deleted that category.
    - Test out `_changelog.md` updates:
      - When there hasn't already been a change this month
      - When a previous change has already been made this month
      - When no new categories have been created, `_changelog` s/b unchanged.
    - When the categories returned from the API exactly match the categories listed in the docs, no changes should be made.
